R/accounts.R

```diff
@@ -41,6 +41,64 @@ accounts <- function(server = NULL) {
   data.frame(name = names, server = servers, stringsAsFactors = FALSE)
 }
 
+#' Connect Api User Account
+#'
+#' Connect a user account to the package using an API key for authentication
+#' so that it can be used to deploy and manage
+#' applications on behalf of the account.
+#'
+#' @param account A name for the account to connect. Optional.
+#' @param server The server to connect to. Optional if there is only one server
+#'   registered.
+#' @param apiKey The API key used to authenticate the user
+#' @param quiet Whether or not to show messages and prompts while connecting the
+#'   account.
+#'
+#' @details This function configures the user to connect using an apiKey in
+#' the http auth headers instead of a token. This is less secure but may
+#' be necessary when the client is behind a proxy or otherwise unable to
+#' authenticate using a token.
+#'
+#' @family Account functions
+#' @export
+connectApiUser <- function(account = NULL, server = NULL, apiKey = NULL, quiet = FALSE) {
+  # if server isn't specified, look up the default
+  if (is.null(server)) {
+    target <- getDefaultServer(local = TRUE)
+  } else {
+    target <- serverInfo(server)
+  }
+
+  if (is.null(target)) {
+    stop("You must specify a server to connect to.")
+  }
+
+  # if account is specified and we already know about the account, get the User
+  # ID so we can prefill authentication fields
+  userId <- 0
+  userAccounts <- accounts(target$name)
+  if (!is.null(account) && !is.null(userAccounts)) {
+    if (account %in% userAccounts[,"name"]) {
+      accountDetails <- accountInfo(account, target$name)
+      userId <- accountDetails$accountId
+      if (!quiet) {
+        message("The account '",  account, "' is already registered; ",
+                "attempting to reconnect it.")
+      }
+    }
+  }
+
+  # write the user info
```

Comment:
I think that prior to writing the user info to disk, we should use the new API key to make a dummy request (best endpoint is probably the current user endpoint used by connectUser). If that request fails, we can tell the user the key is invalid and refuse to register it. That will move most of the connectivity issues from the time when the user attempts to deploy to the time when they register the key.

Reply:
To avoid duping and refactoring shared code for token+API based requests (and _maybe_ adding useful functionality in the `connect.R` client) I have it just grab `server_settings` instead. Theoretically less of a database hit as well. If this isn't sufficient I can change it to grab current user instead.

Reply:
The `server_settings` call does not require auth. If auth is present, it will be used and validated. Is that enough for your purposes?

The "current user" call is used by the IDE+rsconnect as it tries to verify that the IDE<->server auth has been linked.
https://github.com/rstudio/rsconnect/blob/a9fe71bcd8f3c181d6e0c9379a440879400318bb/R/accounts.R#L100-L107

The settings call is used by the IDE+rsconnect as it tries to determine a root URL from a user-entered URL. https://github.com/rstudio/rsconnect/blob/ae203f870f241e7b1e5de35230b635be754785ee/R/ide.R#L11

Reply:
Ah, I had only tested good/bad API keys but not a lack of one. That makes sense. While it does what I intended it's possible that passing in an empty API key would give a false positive. I'll change it back to what I originally had done, thanks!